src/vsr/replica.zig

```diff
@@ -1400,46 +1450,105 @@ pub fn ReplicaType(
             assert(message.header.command == .start_view);
             if (self.ignore_view_change_message(message)) return;
 
-            if (message.header.op > self.op_checkpoint_trigger()) {
-                // This replica is too far behind, i.e. the new `self.op` is too far ahead of the
-                // last checkpoint. If we wrap now, we overwrite un-checkpointed transfers in the WAL,
-                // precluding recovery.
-                //
-                // TODO State transfer. Currently this is unreachable because the
-                // primary won't checkpoint until all replicas are caught up.
-                @panic("unimplemented (state transfer)");
-            }
-
             assert(self.status == .view_change or
                 self.status == .normal or
                 self.status == .recovering_head);
             assert(message.header.view >= self.view);
             assert(message.header.replica != self.replica);
             assert(message.header.replica == self.primary_index(message.header.view));
+            assert(message.header.op >= message.header.commit);
+            assert(message.header.op - message.header.commit <=
+                constants.pipeline_prepare_queue_max);
+
+            if (message.header.view == self.log_view and message.header.op <= self.op) {
+                // We were already in this view prior to receiving the SV.
+                assert(self.status == .normal or self.status == .recovering_head);
+
+                log.debug("{}: on_start_view view={} (ignoring, old message)", .{
+                    self.replica,
+                    self.log_view,
+                });
+                return;
+            }
 
             self.view_jump(message.header);
 
-            assert(self.status == .view_change);
+            assert(self.status == .view_change or self.status == .normal);
```

Comment:
Normal-status (lagging) replicas will now request SV messages for their own view, when they need to advance their head op to some future `commit_max`.